app/internal/astronomy.py

```diff
@@ -0,0 +1,88 @@
+import datetime
+import functools
+import frozendict
+import requests
+from typing import Dict, Tuple, Union
+
+from app import config
+
+
+# This feature requires an API KEY - get yours free @ www.weatherapi.com
+
+SUCCESS_STATUS = 0
+ERROR_STATUS = -1
+ASTRONOMY_URL = "http://api.weatherapi.com/v1/astronomy.json"
+NO_API_RESPONSE = "No response from server"
+
+
+def freezeargs(func):
+    """Transform mutable dictionary into immutable
+    Credit to 'fast_cen' from 'stackoverflow'
+    https://stackoverflow.com/questions/6358481/
+    using-functools-lru-cache-with-dictionary-arguments
+    """
+    @functools.wraps(func)
+    def wrapped(*args, **kwargs):
+        args = tuple([frozendict.frozendict(arg)
+                      if isinstance(arg, dict) else arg for arg in args])
+        kwargs = {k: frozendict.frozendict(v) if isinstance(v, dict) else v
+                  for k, v in kwargs.items()}
+        return func(*args, **kwargs)
+    return wrapped
+
+
+@freezeargs
+@functools.lru_cache(maxsize=128, typed=False)
+def get_data_from_api(formatted_date: str, location: str)\
+        -> Union[Tuple[None, str], Tuple[dict, None]]:
+    """ get the relevant astronomical data by calling the "weather api" API.
+    Args:
+        formatted_date (date) - relevant date.
+        location (str) - location name.
+    Returns:
+        response_json (json dict) - relevant part (data / error) of the
+        JSON returned by the API.
+        error_text (str) - error message.
+    """
+    input_query_string = dict(key=config.ASTRONOMY_API_KEY, q=location,
+                              dt=formatted_date)
+    try:
+        response = requests.request("GET", ASTRONOMY_URL,
```

Comment:
Prefer to work with async library (like httpx)

Reply:
I've changed to asyncio and the code runs fine from the main program,
But the test checks (which runs perfectly on my computer) just fails here:
[https://github.com/PythonFreeCourse/calendar/pull/142/checks?check_run_id=1799100569](url)
Couldn't find and information regarding to what causes the problem...
Any advice?

Reply:
I think that's because you mock using `responses` which is more suitable to `requests` than to `httpx`

Reply:
Thanks. You're right of course. Now I've changed it and it seems that my tests are fine and yet, there is an error in: tests/test_whatsapp.py::test_end_to_end_testing ERROR
[https://github.com/PythonFreeCourse/calendar/pull/142/checks?check_run_id=1809067309](url)
I've no idea on how my code could impact that...
Can you take a look again please?

Reply:
I think it some caveats in the work with async frameworks.

Can you please `return get_data_from_api(formatted_date, location)` instead of putting it in a async loop?

Reply:
Since I use async and httpx I think have to perform the loop - is there another way?
I've tried to do it prior to the return command, but still the whatsapp test fails...